### PR TITLE
Update dependencies, standardize on retrieval via tag where available

### PR DESCRIPTION
**Pull request by moderation**
Signed-off-by: Michael Payne michael@sooper.org

Dependency: Upgrade to backward-cpp v1.4, grpc v1.14.2, googletest v1.8.1, rules_go 0.11.2, buildifier 0.15.0, nghttp2 1.33.0 and subpar 1.3.0. This PR does not update dependencies that rely on commit shas.

Description: nghttp1 1.33.0 [release notes](https://github.com/nghttp2/nghttp2/releases/tag/v1.33.0), backward-cpp v1.4 [release notes](https://github.com/bombela/backward-cpp/releases/tag/v1.4), grpc [release notes](https://github.com/grpc/grpc/releases/tag/v1.14.2), rules_go [release notes](https://github.com/bazelbuild/rules_go/releases/tag/0.11.2), buildifier [release notes](https://github.com/bazelbuild/buildtools/releases/tag/0.15.0), googletest [release notes](https://github.com/google/googletest/releases/tag/release-1.8.1) and subpar [commits](https://github.com/google/subpar/compare/eb23aa7a5361cabc02464476dd080389340a5522...1.3.0) & [release notes](https://github.com/google/subpar/releases). Add commit date comments to libprotobuf-mutator & jwt_verify_lib

Risk Level: Low
Testing: bazel test //test/... and running on local instances.
Docs Changes: none required
Release Notes: none required

### CODE REVIEW

File: bazel/repository_locations.bzl [hunk 1]
```diff
@@ -14,42 +14,40 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "thrift-0.11.0",
     ),
     com_github_bombela_backward = dict(
-        commit = "44ae9609e860e3428cd057f7052e505b4819eb84",  # 2018-02-06
+        commit = "v1.4",
```
Comment (lizan) [inline bazel/repository_locations.bzl]: use `tag` everywhere` for dependencies specified by a tag?

File: bazel/repository_locations.bzl [hunk 2]
```diff
@@ -14,42 +14,40 @@ REPOSITORY_LOCATIONS = dict(
...
     com_github_eile_tclap = dict(
-        commit = "3627d9402e529770df9b0edf2aa8c0e0d6c6bb41",  # tclap-1-2-1-release-final
+        commit = "tclap-1-2-1-release-final",
         remote = "https://github.com/eile/tclap",
     ),
     com_github_fmtlib_fmt = dict(
-        sha256 = "46628a2f068d0e33c716be0ed9dcae4370242df135aed663a180b9fd8e36733d",
-        strip_prefix = "fmt-4.1.0",
-        urls = ["https://github.com/fmtlib/fmt/archive/4.1.0.tar.gz"],
+        commit = "4.1.0",
```
Comment (lizan) [inline bazel/repository_locations.bzl]: What is the motivation moving this from `http_archive` to `git_repository`? (same for spdlog below as well)

**Comment by lizan** [comment]
+1 to having `sha256`. @htuch We might need to switch to http_archive from git_repository, it doesn't seems the Skylark [git_repository](https://github.com/bazelbuild/bazel/blob/0.16.1/tools/build_defs/repo/git.bzl#L154) has sha256 support. It also benefits CI speed.

**Comment by htuch** [comment]
Now that Pandora's box has been opened, should also loop in @jmillikin-stripe to make sure we're all in agreement :)

File: bazel/repository_locations.bzl (hunk 1 above)
Comment (moderation) [reply bazel/repository_locations.bzl]: I tried that but Bazel generates the following error:
```
ERROR: /home/moderation/Library/envoyproxy/envoy/WORKSPACE:6:1: Traceback (most recent call last):
        File "/home/moderation/Library/envoyproxy/envoy/WORKSPACE", line 6
                envoy_dependencies()
        File "/home/moderation/Library/envoyproxy/envoy/bazel/repositories.bzl", line 286, in envoy_dependencies
                _com_github_bombela_backward()
        File "/home/moderation/Library/envoyproxy/envoy/bazel/repositories.bzl", line 320, in _com_github_bombela_backward
                _repository_impl(name = "com_github_bombela_backw...", ...")
        File "/home/moderation/Library/envoyproxy/envoy/bazel/repositories.bzl", line 36, in _repository_impl
                fail(("Refusing to depend on Git tag ...)))
Refusing to depend on Git tag "v1.4" for external dependency "com_github_bombela_backward": use 'commit' instead.
ERROR: Error evaluating WORKSPACE file
ERROR: error loading package '': Encountered error while reading extension file 'bazel/repositories.bzl': no such package '@envoy_api//bazel': error loading package 'external': Could not load //external package
```

File: bazel/repository_locations.bzl (hunk 2 above)
Comment (moderation) [reply bazel/repository_locations.bzl]: Consistency was the only motivator.

**Comment by moderation** [comment]
I have a commit that add the sha256 for each of the dependencies. What is the impact of  https://docs.bazel.build/versions/master/be/workspace.html#git_repository.sha256 that says
> Specifying this forces the repository to be downloaded as a tarball.

File: ci/build_container/build_recipes/nghttp2.sh [hunk 3]
```diff
@@ -2,12 +2,9 @@
...
 set -e
 
-# Use master branch, which contains a fix for the spurious limit of 100 concurrent streams:
-# https://github.com/nghttp2/nghttp2/commit/2ba1389993729fcb6ee5794ac512f2b67b29952e
-# TODO(PiotrSikora): switch back to releases once v1.33.0 is out.
-VERSION=e5b3f9addd49bca27e2f99c5c65a564eb5c0cf6d  # 2018-06-09
+VERSION=1.33.0
 
-curl https://github.com/nghttp2/nghttp2/archive/"$VERSION".tar.gz -sLo nghttp2-"$VERSION".tar.gz
+curl https://github.com/nghttp2/nghttp2/releases/download/v"$VERSION"/nghttp2-"$VERSION".tar.gz -sLo nghttp2-"$VERSION".tar.gz
```
Comment (htuch) [inline ci/build_container/build_recipes/nghttp2.sh]: Could we add a SHA256 verification step here? Should be a quick one liner.

**Comment by lizan** [comment]
LGTM, thanks!

File: ci/build_container/build_recipes/nghttp2.sh (hunk 3 above)
Comment (moderation) [reply ci/build_container/build_recipes/nghttp2.sh]: We aren't sha256 verifying anything in `ci/build_container/build_recipes/*`. I think I know how to do this - will report back.

File: ci/build_container/build_recipes/zlib.sh [hunk 4]
```diff
@@ -4,11 +4,14 @@ set -e
 
 VERSION=1.2.11
 
-curl https://github.com/madler/zlib/archive/v"$VERSION".tar.gz -sLo zlib-"$VERSION".tar.gz
+curl https://github.com/madler/zlib/archive/v"$VERSION".tar.gz -sLo zlib-"$VERSION".tar.gz \
+  && echo '629380c90a77b964d896ed37163f5c3a34f6e6d897311f1df2a7016355c45eff' zlib-"$VERSION".tar.gz | sha256sum --check
```
Comment (htuch) [inline ci/build_container/build_recipes/zlib.sh]: Can you restructure to make the SHA a variable like the VERSION above?

**Comment by htuch** [comment]
@zuercher looks like we're missing `sha256sum` in the CI image for CircleCI, this is why the Mac build fails. Thoughts? I suppose we should add this as a dependency in https://github.com/envoyproxy/envoy/tree/master/bazel#quick-start-bazel-build-for-developers @moderation.

File: ci/build_container/build_recipes/nghttp2.sh (hunk 3 above)
Comment (moderation) [reply ci/build_container/build_recipes/nghttp2.sh]: @htuch the commit to add sha256 verification is at https://github.com/envoyproxy/envoy/pull/4353/commits/28f0f5867c380663ec6367061b6fd06d6e2b6818 but the change is failing for the Mac build as `sha256sum` doesn't exist. I'll do some testing with the solutions at https://github.com/ESGF/esg-search/issues/84

File: bazel/repository_locations.bzl [hunk 5]
```diff
@@ -14,42 +14,47 @@ REPOSITORY_LOCATIONS = dict(
         strip_prefix = "thrift-0.11.0",
     ),
     com_github_bombela_backward = dict(
-        commit = "44ae9609e860e3428cd057f7052e505b4819eb84",  # 2018-02-06
+        sha256 = "ad73be31c5cfcbffbde7d34dba18158a42043a109e7f41946f0b0abd589ed55e",
+        commit = "v1.4",
```
Comment (jmillikin-stripe) [inline bazel/repository_locations.bzl]: I'm not sure exactly what this is trying to express, but it won't do what you want.

1. Setting a SHA256 can only work when downloading a single file, such as a tarball. Git clones don't have a deterministic sha256 so there's nowhere the `sha256` value could be used or validated.

2. The `commit` kwarg to `git_repository` is expected to be a Git commit ID. For fetching by tag, you want the `tag` kwarg. Note that this is usually considered less secure + deterministic than fetching by commit, because Git tags can be arbitrarily changed by the repository owner.

Comment (htuch) [reply bazel/repository_locations.bzl]: I think tag is nicer if we can get a SHA256 for verification; we get self-documenting specs and a stronger hash than that provided by native git.

Comment (jmillikin-stripe) [reply bazel/repository_locations.bzl]: What byte blob is the sha256 calculated from?

Comment (moderation) [reply bazel/repository_locations.bzl]: See prior comments https://github.com/envoyproxy/envoy/pull/4353#issuecomment-419088327 and  https://github.com/envoyproxy/envoy/pull/4353/commits/fe04e3913f8294931c41b78504608e302691748f#r215618035. kwarg of `tag` generates an error. By specifying the sha256 we are downloading the tar file.

Comment (jmillikin-stripe) [reply bazel/repository_locations.bzl]: Are you sure? Envoy's repositories.bzl uses the Skylark implementation of `git_repository` from https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/git.bzl, which doesn't have any hits for "sha256" or "github".

Comment (moderation) [reply bazel/repository_locations.bzl]: I think @jmillikin-stripe is right. Adding an invalid sha passes through without error in `bazel/repository_locations.bzl`. Passing in `fakevalue` instead of `sha256` works too so Bazel is ignoring what it doesn't comprehend. 

My guess is that https://github.com/envoyproxy/envoy/blob/master/bazel/repository_locations.bzl#L33-L35 works as it invokes https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/http.bzl which does support `sha256` - https://github.com/bazelbuild/bazel/blob/master/tools/build_defs/repo/http.bzl#L122

In summary, using git means we can't check the sha. We would need to move away from tags to commit ID AND use the http method. 

@htuch the original vision of greater maintainability with sha256 checked security isn't panning out :-(

Comment (jmillikin-stripe) [reply bazel/repository_locations.bzl]: If you're willing to have a stricter dependency on Github and a slightly risk of build breakage when they change their tarball builder, you could use Github's auto-generated tarballs with sha256 attrs.

Comment (htuch) [reply bazel/repository_locations.bzl]: @jmillikin-stripe do you think GH have reduced their churn? I remember the original concern due to GH breaking arbitrarily URLs and SHAs, has this stabilized since? If so, that seems entirely desirable.

File: bazel/repository_locations.bzl [hunk 6]
```diff
@@ -10,24 +10,27 @@ REPOSITORY_LOCATIONS = dict(
...
     com_github_apache_thrift = dict(
         sha256 = "7d59ac4fdcb2c58037ebd4a9da5f9a49e3e034bf75b3f26d9fe48ba3d8806e6b",
-        urls = ["https://files.pythonhosted.org/packages/c6/b4/510617906f8e0c5660e7d96fbc5585113f83ad547a3989b80297ac72a74c/thrift-0.11.0.tar.gz"],  # 0.11.0
         strip_prefix = "thrift-0.11.0",
+        urls = ["https://files.pythonhosted.org/packages/c6/b4/510617906f8e0c5660e7d96fbc5585113f83ad547a3989b80297ac72a74c/thrift-0.11.0.tar.gz"],
     ),
     com_github_bombela_backward = dict(
-        commit = "44ae9609e860e3428cd057f7052e505b4819eb84",  # 2018-02-06
-        remote = "https://github.com/bombela/backward-cpp",
+        sha256 = "ad73be31c5cfcbffbde7d34dba18158a42043a109e7f41946f0b0abd589ed55e",
```
Comment (htuch) [inline bazel/repository_locations.bzl]: Super awesome, I think even without tags this is much clearer. One quick check; did you verify the `sha256` validation is working here as intended?

Comment (moderation) [reply bazel/repository_locations.bzl]: I did check the validation and it is working. I changed a bunch of sha's to validate and they fail on build as intended. I also had an incorrect sha for gcovr and pushed a commit to fix. So I think this is ready.
